clash-cores/test/Test/Cores/Xilinx/DcFifo.hs

```diff
@@ -0,0 +1,92 @@
+{-# LANGUAGE OverloadedStrings #-}
+
+module Test.Cores.Xilinx.DcFifo ( tests ) where
+
+import qualified Prelude as P
+import qualified Data.List as L
+import Data.Maybe (catMaybes)
+import Control.Monad (replicateM)
+
+import Clash.Explicit.Prelude
+import Clash.Cores.Xilinx.DcFifo.Explicit
+
+import qualified Hedgehog.Gen as Gen
+import qualified Hedgehog.Range as Range
+import Hedgehog (MonadGen, (===), property, forAll, Property)
+import Test.Tasty.Hedgehog (testPropertyNamed)
+import Test.Tasty (TestTree)
+
+tests :: TestTree
+tests = testPropertyNamed
+  "FIFO doesn't lose any data"
+  "prop_fifo"
+  prop_fifo
+
+intersperseStalls ::
+  -- | Data
+  [a] ->
+  -- | Stalls
+  [a] ->
+  [a]
+intersperseStalls (d:ds) s = d:s P.++ intersperseStalls ds s
+intersperseStalls [] _ = []
+
+prop_fifo :: Property
+prop_fifo = property $ do
+  xs <- fmap Just <$> forAll (replicateM 10 genData)
+  stallRead <- forAll (Gen.maybe (Gen.int (Range.linear 0 10)))
+  stallWrite <- forAll (Gen.maybe (Gen.int (Range.linear 0 10)))
+  let iWrites = case stallRead of
+        Nothing -> []
+        Just i -> P.replicate i Nothing
+  let readStalls = case stallWrite of
+        Nothing -> [False]
+        Just i -> False : L.replicate i True
+  throughFifo (intersperseStalls xs iWrites) (cycle readStalls) === catMaybes xs
+
+
+genData :: (KnownNat n, MonadGen m) => m (BitVector n)
+genData = Gen.enumBounded
+
+takeState ::
+  (Bool, [Bool]) ->
+  (ResetBusy, Empty, DataCount depth, BitVector n) ->
+  ((Bool, [Bool]), (Maybe (BitVector n), Bool))
+takeState (_,[]) _ = error "Unreachable."
+takeState (_, _:stalls) (1, _, _, _) = ((False, stalls), (Nothing, False))
+takeState (readLastCycle, True:stalls) (_, _, _, d) = ((False, stalls), (nextData, False))
+  where
+    nextData = if readLastCycle then Just d else Nothing
```

Comment:
```suggestion
    nextData = readLastCycle `orNothing` d
```

I see this pattern a lot, so I wrote a little helper function for it. You can decide for yourself if you want to to use it :)